### PR TITLE
Add documentation on instructions module for Bootcamp website

### PR DESCRIPTION
**Pull request by aakahlow**

### CODE REVIEW

File: modules/Developing gem5 models/_posts/2000-07-01-instructions.md [hunk 1]
```diff
@@ -1,6 +1,132 @@
...
+
+Moreover, the templates are usually specialized by passing an object of `InstObjParams` class to them. `InstObjParams` encapsulates the full set of parameters to be substituted into a template. It is also important to note that, a developer would specify or define some symbols that are used inside the templates, but the ISA parser would also define multiple other symbols that are then used inside the templates. ISA parser defines these symbols by inferring certain information about the instruction from other components like instruction formats.
+Below are a few examples of template blocks and the corresponding C++ code that will be generated using them.
+
+![ROp format used by R format instructions](../../../assets/img/ROpFormat.png)
```
Comment (BobbyRBruce) [inline modules/Developing gem5 models/_posts/2000-07-01-instructions.md:106]: Fix link

File: modules/Developing gem5 models/_posts/2000-07-01-instructions.md [hunk 2]
```diff
@@ -1,6 +1,132 @@
...
+*ROp format is used by RISC-V add instruction. Different templates are specialized using InstObjParams object to generate the output blocks.*
+
+![Example of BasicDeclare template](../../../assets/img/BasicDeclare.png)
```
Comment (BobbyRBruce) [inline modules/Developing gem5 models/_posts/2000-07-01-instructions.md:109]: Fix link

File: modules/Developing gem5 models/_posts/2000-07-01-instructions.md [hunk 3]
```diff
@@ -1,6 +1,132 @@
...
+*BasicDeclare template is specialized to generate the output that goes in the header files. Above picture shows how the DSL template is converted into C++ code for an add instruction.*
+
+![Example of BasicExecute template](../../../assets/img/ExecuteTemplate.png)
```
Comment (BobbyRBruce) [inline modules/Developing gem5 models/_posts/2000-07-01-instructions.md:112]: Fix link

**Review by BobbyRBruce — CHANGES_REQUESTED** [review]
This is really good. Just fix up the links and i'll approve it.

File: modules/Developing gem5 models/_posts/2000-07-01-instructions.md [hunk 4]
```diff
@@ -1,6 +1,132 @@
...
+
+To be able to use any ISA with any CPU model in gem5, it tries to keep the two independent of each other.
+gem5 relies on two generic interfaces to maintain this independence:
+
+- Static instruction: CPU models access the ISA sub-system via static instruction objects.
+- Execution context: ISA sub-system accesses the CPU model via execution context interface.
+
+Following picture provides a detailed view of what gem5 components are ISA independent, what are ISA dependent, and what are in between.
+
+![ISA dependent and independent components of gem5](../../../assets/img/ISAInd.png)
```
[the remark this answers is not in the document]
Comment (aakahlow) [reply modules/Developing gem5 models/_posts/2000-07-01-instructions.md]: Done

File: modules/Developing gem5 models/_posts/2000-07-01-instructions.md [hunk 5]
```diff
@@ -1,6 +1,132 @@
...
+Next, the idea is to run the same program with `gdb` and to put some breakpoints in the gem5 source code (at the important functions that are defined in the ISA sub-system and are responsible for the decoding and the execution of instructions).
+Instructions in slide 19, can be used to put breakpoints on relevant functions in a RISC-V Add and Lw (load word) instruction.By printing the backtrace of the previous function calls at the breakpoint we can observe which CPU functions are responsible for calling the ISA-subsystem functions.
+As we will see in the next section, the functions where we put our breakpoints are not handwritten code, rather they are auto-generated via the gem5 ISA parser.
+
+## gem5 ISA parser
+
+gem5 uses a DSL (domain specific language) to specify the encoding and execution functions of instructions.
+This DSL is understandable by a gem5 ISA parser, which might infer extra information from the ISA description file and finally generates the C++ code which is then compiled into the gem5 binary. Following picture provides a high-level instruction definition flow in gem5.
+
+![High level flow of instruction definition in gem5](../../../assets/img/ISAParser.png)
```
[the remark this answers is not in the document]
Comment (aakahlow) [reply modules/Developing gem5 models/_posts/2000-07-01-instructions.md]: fixed

File: modules/Developing gem5 models/_posts/2000-07-01-instructions.md [hunk 6]
```diff
@@ -1,6 +1,132 @@
...
+
+Slides (slide 24 -- 88) provide a detailed explanation of gem5 ISA specification DSL. Following is a summary of the details in those slides.
+
+ISA specification files contain different kind of sections. For example, decode section is one of those sections and is responsible for specifying how an instruction would be decoded.
+The decode blocks look like a C switch statements block. The decode block can be nested and usually rely on multiple fields in the instruction to determine what instruction is under consideration. Below is an example of a decode block which is responsible for decoding a RISC-V Add instruction and the corresponding C++ code that is generated by the ISA parser.
+
+![Decode block in gem5 ISA DSL](../../../assets/img/decode.png)
```
[the remark this answers is not in the document]
Comment (aakahlow) [reply modules/Developing gem5 models/_posts/2000-07-01-instructions.md]: fixed

File: modules/Developing gem5 models/_posts/2000-07-01-instructions.md (hunk 1 above)
Comment (aakahlow) [reply modules/Developing gem5 models/_posts/2000-07-01-instructions.md]: fixed

File: modules/Developing gem5 models/_posts/2000-07-01-instructions.md (hunk 2 above)
Comment (aakahlow) [reply modules/Developing gem5 models/_posts/2000-07-01-instructions.md]: fixed

File: modules/Developing gem5 models/_posts/2000-07-01-instructions.md (hunk 3 above)
Comment (aakahlow) [reply modules/Developing gem5 models/_posts/2000-07-01-instructions.md]: fixed

**Review by BobbyRBruce — APPROVED** [review]

**Pull request merged**